### PR TITLE
BUG: memmap close files when it shouldn't, load leaves them open when it shouldn't

### PR DESCRIPTION
**Pull request by eric-wieser**
This would be straightforward with `ExitStack`, but I don't feel like backporting something that big from py3k.

### CODE REVIEW

**Comment by eric-wieser** [comment]
The diff in `memmap` isn't super clear because the level of indentation changed. In short:

* When passed a raw file handle, wrap it in a null context that does nothing
* Do the cheap checks before opening the file

File: numpy/compat/py3k.py [hunk 1]
```diff
@@ -97,6 +97,28 @@ def is_pathlib_path(obj):
     """
     return Path is not None and isinstance(obj, Path)
 
+# from Python 3.7
+class contextlib_nullcontext(object):
```
Comment (mhvk) [inline numpy/compat/py3k.py]: This should go into `__all__`

File: numpy/core/memmap.py [hunk 2]
```diff
@@ -211,78 +213,72 @@ def __new__(subtype, filename, dtype=uint8, mode='r+', offset=0,
...
+        if hasattr(filename, 'read'):
+            f_ctx = contextlib_nullcontext(filename)
         elif is_pathlib_path(filename):
-            self.filename = filename.resolve()
-        # py3 returns int for TemporaryFile().name
-        elif (hasattr(filename, "name") and
-              isinstance(filename.name, basestring)):
-            self.filename = os.path.abspath(filename.name)
-        # same as memmap copies (e.g. memmap + 1)
+            f_ctx = filename.open((mode == 'c' and 'r' or mode)+'b')
```
Comment (mhvk) [inline numpy/core/memmap.py]: While you are at it: `('r' if mode == 'c' else mode) + 'b'`

File: numpy/core/memmap.py [hunk 3]
```diff
@@ -211,78 +213,72 @@ def __new__(subtype, filename, dtype=uint8, mode='r+', offset=0,
...
+            fid.seek(0, 2)
+            flen = fid.tell()
+            descr = dtypedescr(dtype)
+            _dbytes = descr.itemsize
+
+            if shape is None:
+                bytes = flen - offset
+                if bytes % _dbytes:
+                    raise ValueError("Size of available data is not a "
+                            "multiple of the data-type size.")
```
Comment (mhvk) [inline numpy/core/memmap.py]: Strange indentation.

**Review by mhvk — COMMENTED** [review]
Looks good. Failures are just due to not adding the new null context manager to `__all__`

**Comment by mhvk** [comment]
Want to wrap this up? From my comments, it seems there was only a trivial failure and an even more trivial nitpick to address.

**Comment by eric-wieser** [comment]
`__all__` entry added, and conditional nit addressed. Didn't mess with the formatting, because that's not part of this patch anyway

**Review by mhvk — APPROVED** [review]
Fine with ignoring the alignment for now.

**Pull request merged**